web/core/src/replay-visualizer-factory.ts

```diff
@@ -0,0 +1,93 @@
+import { GameAdapter } from './adapter';
+import { ReplayVisualizer } from './player';
+
+/**
+ * The shape of the persistent state for a single ReplayVisualizer instance.
+ */
+interface ReplayVisualizerState {
+  replay: any | null;
+  agents: any[];
+  step: number;
+  controls: boolean;
+  playing: boolean;
+}
+
+// --- HMR State Management ---
+let hmrReplayVisualizerStates: Map<string, ReplayVisualizerState>;
+let hmrReplayVisualizerInstances: Map<string, ReplayVisualizer>;
+
+if (import.meta.env?.DEV && import.meta.hot) {
+  // 2. Initialize or retrieve the persistent state Map
+  if (!import.meta.hot.data.replayVisualizerStates) {
+    import.meta.hot.data.replayVisualizerStates = new Map<string, ReplayVisualizerState>();
```

Comment:
I haven't seen this `import.` syntax before- would you mind either giving a one sentence summary of what's going on with that or drop a link to somewhere I could learn more?

Reply:
It's vite magic - vite injects this stuff in dev mode https://vite.dev/guide/env-and-mode